### PR TITLE
namespace with a v1 folder

### PR DESCRIPTION
**Pull request by BenMorganIO**
Conflicts:
    api/app/controllers/spree/api/v1/config_controller.rb
    api/config/routes.rb

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://benmorganio-api-version-routes.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://benmorganio-api-version-routes.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by BenMorganIO** [comment]
All of these commits had merge conflicts. I'm thinking that was the reason they didn't make their way in. It was a big bang PR, so I guess this could be expected.

I think that's it. I'll just update for hound and hope to get this merged ASAP.

File: api/app/controllers/spree/api/base_controller.rb [hunk 1]
```diff
@@ -2,170 +2,177 @@
...
-        end
+      def requires_authentication?
+        Spree::Api::Config[:requires_authentication]
+      end
 
-        def current_ability
-          Spree::Ability.new(current_api_user)
-        end
+      def not_found
+        render "spree/api/errors/not_found", status: 404 and return
```
Comment (BenMorganIO) [inline api/app/controllers/spree/api/base_controller.rb]: ignored

File: api/app/controllers/spree/api/base_controller.rb [hunk 2]
```diff
@@ -2,170 +2,177 @@
...
+        else
+          []
         end
+      end
 
-        def error_during_processing(exception)
-          Rails.logger.error exception.message
-          Rails.logger.error exception.backtrace.join("\n")
+      def unauthorized
+        render "spree/api/errors/unauthorized", status: 401 and return
```
Comment (BenMorganIO) [inline api/app/controllers/spree/api/base_controller.rb]: ignored

File: api/app/controllers/spree/api/base_controller.rb [hunk 3]
```diff
@@ -2,170 +2,180 @@
...
+      def load_user
+        @current_api_user = (try_spree_current_user || Spree.user_class.find_by(spree_api_key: api_key.to_s))
+      end
+
+      def authenticate_user
+        unless @current_api_user
+          if requires_authentication? && api_key.blank? && order_token.blank?
+            render "spree/api/errors/must_specify_api_key", status: 401 and return
+          elsif order_token.blank? && (requires_authentication? || api_key.present?)
+            render "spree/api/errors/invalid_api_key", status: 401 and return
```
Comment (BenMorganIO) [inline api/app/controllers/spree/api/base_controller.rb]: ignored

**Comment by JDutil** [comment]
Part of the issue was that the work wasn't rebased with master, and this still isn't rebased with master.  This undoes the security changes we made yesterday.

File: api/app/controllers/spree/api/base_controller.rb [hunk 4]
```diff
@@ -2,170 +2,176 @@
...
-            "text/xml; charset=utf-8"
-          end
-          headers["Content-Type"] = content_type
+      before_action :set_content_type
+      before_action :load_user
+      before_action :authorize_for_order, if: Proc.new { order_token.present? }
+      before_action :authenticate_user
+      before_action :load_user_roles
+
+      after_filter  :set_jsonp_format
```
Comment (JDutil) [inline api/app/controllers/spree/api/base_controller.rb]: Remove this line.

File: api/app/controllers/spree/api/base_controller.rb [hunk 5]
```diff
@@ -2,170 +2,176 @@
...
+
+      rescue_from Exception, with: :error_during_processing
+      rescue_from ActiveRecord::RecordNotFound, with: :not_found
+      rescue_from CanCan::AccessDenied, with: :unauthorized
+      rescue_from Spree::Core::GatewayError, with: :gateway_error
+
+      helper Spree::Api::ApiHelpers
+
+      def set_jsonp_format
```
Comment (JDutil) [inline api/app/controllers/spree/api/base_controller.rb]: remove this method

File: api/app/controllers/spree/api/base_controller.rb [hunk 6]
```diff
@@ -2,170 +2,180 @@
...
+
+      def set_content_type
+        headers["Content-Type"] = content_type
+      end
 
-        def load_user_roles
-          @current_user_roles = if @current_api_user
-            @current_api_user.spree_roles.pluck(:name)
+      def load_user
+        @current_api_user = (try_spree_current_user || Spree.user_class.find_by(spree_api_key: api_key.to_s))
```
Comment (JDutil) [inline api/app/controllers/spree/api/base_controller.rb]: remove try_spree_current_user we only lookup by api key now

**Comment by BenMorganIO** [comment]
I'll cherry pick the security changes in. See what happens.

**Comment by JDutil** [comment]
Okay merged other PR so you can rebase this.

**Comment by BenMorganIO** [comment]
OK, lets see how CircleCI goes and merge this baby.

**Pull request closed without merging**